File: bin/nodes/sources/postgres_table.dart
import 'dart:async';

import 'package:postgres/postgres.dart';
import 'package:realtime_client/realtime_client.dart';

import 'list_source.dart';

typedef PostgresRowMap = Map<String, Map<String, dynamic>>;

class PostgresTableSource extends ListSource<PostgresRowMap> {
  final PostgreSQLConnection connection;
  final String tableName;

  StreamController<List<PostgresRowMap>>? _queryStreamController;
  RealtimeClient? _realtimeSocket;

  PostgresTableSource(this.connection, this.tableName);

  @override
  Future<void> init() async {
    await super.init();
    setRows(await _fetchResults());

    // await setupPolling();
    await setupRealtime();
  }

  Future<List<PostgresRowMap>> _fetchResults() async =>
      connection.mappedResultsQuery(
        "SELECT * FROM $tableName",
      );

  setupPolling() async {
    _queryStreamController = StreamController();
    _queryStreamController?.addStream(
      Stream.periodic(Duration(seconds: 1)).asyncMap(
        (_) => _fetchResults(),
      ),
    );
    _queryStreamController?.stream.listen(setRows);
  }

  setupRealtime() async {
    final realtimeSocket = RealtimeClient(
      'ws://localhost:4000/socket',
      logger: (kind, msg, data) => print('$kind $msg $data'),
    );

    _realtimeSocket = _realtimeSocket;

    final channel = realtimeSocket.channel('realtime:public:$tableName');
    channel.on(
      'INSERT',
      (payload, {ref}) => insertRow({
        tableName: convertChangeData(
          (payload['columns'] as List).cast<Map<String, dynamic>>(),
          payload['record'],
        ),
      }),
    );
    channel.on('UPDATE', (payload, {ref}) {
      final typedRow = {
        tableName: convertChangeData(
          (payload['columns'] as List).cast<Map<String, dynamic>>(),
          payload['record'],
        )
      };
      updateRows(
        (row) => row[tableName]!['id'] == typedRow[tableName]!['id']
            ? typedRow
            : row,
      );
    });

    realtimeSocket.connect();
    channel.subscribe();

    // Wait for channel to be joined
    int maxRetries = 5;
    while (!channel.isJoined() && maxRetries-- > 0)
      await Future.delayed(Duration(milliseconds: 200));
  }

  @override
  Future close() async {
    await super.close();
    await _queryStreamController?.close();
    _realtimeSocket?.disconnect();
  }
}

File: src/nodes/sources/postgres_table.ts
import { RealtimeClient, RealtimeSubscription } from "@supabase/realtime-js"
import { convertChangeData } from "@supabase/realtime-js/dist/main/lib/transformers"
import { appendAsyncConstructor } from "async-constructor"
import _ from "lodash"
import { Client } from 'pg'
import { delayed } from "../../utils"
import { ArraySource } from "./array"

/**
 * Ouputs the rows of a postgres table.
 * 
 * This data is eventually up to date as data changes in the table.
 * Each time some data is added, modified or deleted from the table
 * this node re-emits the up-to-date rows.
 * One of two mechanism can be used, polling (every second) or realtime (needs
 * a supabase-realtime client connected to a supabse-realtime server).
 */
export class PostgresTableSource extends ArraySource<Object> {
    readonly postgresClient: Client
    readonly realtimeClient?: RealtimeClient
    readonly tableName: string
    private realtimeChannel?: RealtimeSubscription

    constructor(postgresClient: Client, tableName: string, realtimeClient?: RealtimeClient) {
        super()
        this.postgresClient = postgresClient
        this.tableName = tableName
        this.realtimeClient = realtimeClient
        this.nodeId = `${this.nodeId}(${this.tableName})`
        appendAsyncConstructor(this, async () => {
            await this.fetchResults()
            if (_.isNil(this.realtimeClient))
                this.setupPolling()
            else
                await this.setupRealtime()
        })
    }

    async close() {
        this.realtimeChannel?.unsubscribe()
        await super.close()
    }

    private async fetchResults() {
        const { rows } = await this.postgresClient.query(
            `SELECT * FROM ${this.tableName}`,
        )
        this.setRows(rows)
    }

    private setupPolling() {
        const timer = setInterval(() => {
            if (this.isClosed)
                clearInterval(timer)
            else
                this.fetchResults()
        }, 1000)
    }

    private async setupRealtime() {
        if (_.isNil(this.realtimeClient))
            throw Error("Can't setup realtime without a realtime client")

        const channel = this.realtimeClient.channel(`realtime:public:${this.tableName}`)
        channel.on(
            'INSERT',
            ({ columns, record }: { columns: any, record: any }) =>
                this.insertRow(
                    convertChangeData(columns, record)
                ),
        )
        channel.on(
            'UPDATE',
            ({ columns, record, old_record }: { columns: any, record: any, old_record: any }) => {
                const typedNewRecord = convertChangeData(columns, record)
                const typedOldRecord = convertChangeData(columns, old_record)
                this.updateRows((row: any) =>
                    _.isEqual(row, typedOldRecord) ? typedNewRecord : row,
                )
            },
        )
        channel.on(
            'DELETE',
            ({ columns, old_record }: { columns: any, record: any, old_record: any }) => {
                const typedOldRecord = convertChangeData(columns, old_record)
                let hasDeletedARow = false
                this.updateRows((row: any) => {
                    if (!hasDeletedARow && _.isEqual(typedOldRecord, row)) {
                        hasDeletedARow = true
                        return null
                    } else
                        return row
                })
            },
        )
        channel.subscribe()
        this.realtimeChannel = channel

        // Wait for channel to be joined
        let maxRetries = 10
        while (!channel.isJoined() && maxRetries-- > 0)
            await delayed(50, () => { })
    }
}